File: packages/prebid_mobile_sdk_max/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_max/MaxBannerAdViewFactory.kt
package com.prebid.prebid_mobile_sdk_max

import android.app.Activity
import android.content.Context
import android.view.View
import com.applovin.mediation.MaxAd
import com.applovin.mediation.MaxAdViewAdListener
import com.applovin.mediation.MaxError
import com.applovin.mediation.adapters.prebid.utils.MaxMediationBannerUtils
import com.applovin.mediation.ads.MaxAdView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.prebid.mobile.AdSize
import org.prebid.mobile.api.mediation.MediationBannerAdUnit

/// PlatformView factory for AppLovin MAX-mediated banners. The rendered view is
/// the MAX [MaxAdView]; Prebid's [MediationBannerAdUnit] runs the auction and
/// passes the winning bid to MAX via the Prebid MAX adapter.
class MaxBannerAdViewFactory(
    private val messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<*, *> ?: emptyMap<String, Any>()
        return MaxBannerPlatformView(
            activityProvider() ?: context,
            viewId,
            messenger,
            params,
        )
    }
}

class MaxBannerPlatformView(
    context: Context,
    viewId: Int,
    messenger: BinaryMessenger,
    params: Map<*, *>,
) : PlatformView {

    private val adView: MaxAdView
    private val methodChannel: MethodChannel
    private var adUnit: MediationBannerAdUnit? = null

    init {
        val configId = params["configId"] as? String ?: ""
        val maxAdUnitId = params["maxAdUnitId"] as? String ?: ""
        val width = params["width"] as? Int ?: 320
        val height = params["height"] as? Int ?: 50
        val autoLoad = params["autoLoad"] as? Boolean ?: true

        methodChannel = MethodChannel(messenger, "prebid_mobile_sdk_max/banner_$viewId")

        adView = MaxAdView(maxAdUnitId, context)
        adView.setListener(object : MaxAdViewAdListener {
            override fun onAdLoaded(ad: MaxAd) {
                methodChannel.invokeMethod(
                    "onAdSize",
                    mapOf("width" to width.toDouble(), "height" to height.toDouble()),
                )
                methodChannel.invokeMethod("onAdLoaded", null)
                methodChannel.invokeMethod("onAdDisplayed", null)
            }

            override fun onAdLoadFailed(adUnitId: String, error: MaxError) {
                methodChannel.invokeMethod("onAdFailed", error.message)
            }

            override fun onAdDisplayFailed(ad: MaxAd, error: MaxError) {
                methodChannel.invokeMethod("onAdFailed", error.message)
            }

            override fun onAdClicked(ad: MaxAd) {
                methodChannel.invokeMethod("onAdClicked", null)
            }

            override fun onAdHidden(ad: MaxAd) {
                methodChannel.invokeMethod("onAdClosed", null)
            }

            override fun onAdDisplayed(ad: MaxAd) {}
            override fun onAdExpanded(ad: MaxAd) {}
            override fun onAdCollapsed(ad: MaxAd) {}
        })

        val mediationUtils = MaxMediationBannerUtils(adView)
        adUnit = MediationBannerAdUnit(
            context,
            configId,
            AdSize(width, height),
            mediationUtils,
        )

        if (autoLoad) {
            adUnit?.fetchDemand {
                adView.loadAd()
            }
        }
    }

    override fun getView(): View = adView

    override fun dispose() {
        adUnit?.destroy()
        adUnit = null
        adView.destroy()
    }
}

File: packages/prebid_mobile_sdk_max/ios/prebid_mobile_sdk_max/Sources/prebid_mobile_sdk_max/MaxBannerAdViewFactory.swift
import Flutter
import UIKit
import PrebidMobile
import PrebidMobileMAXAdapters
import AppLovinSDK

/// PlatformView factory for AppLovin MAX-mediated banners. The rendered view is
/// the MAX `MAAdView`; Prebid's `MediationBannerAdUnit` runs the auction and
/// passes the winning bid to MAX via the Prebid MAX adapter.
class MaxBannerAdViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return MaxBannerPlatformView(
            frame: frame,
            viewId: viewId,
            messenger: messenger,
            args: args as? [String: Any] ?? [:]
        )
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class MaxBannerPlatformView: NSObject, FlutterPlatformView, MAAdViewAdDelegate {

    private let maxAdBannerView: MAAdView
    private let methodChannel: FlutterMethodChannel
    private let adSize: CGSize

    // Retained for the lifetime of the view — the auction runs through these.
    private var mediationDelegate: MAXMediationBannerUtils?
    private var mediationAdUnit: MediationBannerAdUnit?

    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        args: [String: Any]
    ) {
        let configId = args["configId"] as? String ?? ""
        let maxAdUnitId = args["maxAdUnitId"] as? String ?? ""
        let width = args["width"] as? Int ?? 320
        let height = args["height"] as? Int ?? 50
        let autoLoad = args["autoLoad"] as? Bool ?? true

        adSize = CGSize(width: width, height: height)

        methodChannel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_max/banner_\(viewId)",
            binaryMessenger: messenger
        )

        // 1. Create and configure the MAX ad view.
        maxAdBannerView = MAAdView(adUnitIdentifier: maxAdUnitId)
        maxAdBannerView.frame = CGRect(origin: .zero, size: adSize)
        maxAdBannerView.isHidden = false

        super.init()

        maxAdBannerView.delegate = self

        // 2. Prebid mediation utils + ad unit.
        let mediationDelegate = MAXMediationBannerUtils(adView: maxAdBannerView)
        self.mediationDelegate = mediationDelegate
        let adUnit = MediationBannerAdUnit(
            configID: configId,
            size: adSize,
            mediationDelegate: mediationDelegate
        )
        mediationAdUnit = adUnit

        if autoLoad {
            // 3. Fetch demand, then load the MAX banner.
            adUnit.fetchDemand { [weak self] _ in
                self?.maxAdBannerView.loadAd()
            }
        }
    }

    func view() -> UIView {
        return maxAdBannerView
    }

    // MARK: - MAAdViewAdDelegate

    func didLoad(_ ad: MAAd) {
        methodChannel.invokeMethod("onAdSize", arguments: [
            "width": Double(adSize.width),
            "height": Double(adSize.height),
        ])
        methodChannel.invokeMethod("onAdLoaded", arguments: nil)
        methodChannel.invokeMethod("onAdDisplayed", arguments: nil)
    }

    func didFailToLoadAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {
        let nsError = NSError(
            domain: "MAX",
            code: error.code.rawValue,
            userInfo: [NSLocalizedDescriptionKey: error.message]
        )
        mediationAdUnit?.adObjectDidFailToLoadAd(adObject: maxAdBannerView, with: nsError)
        methodChannel.invokeMethod("onAdFailed", arguments: error.message)
    }

    func didFail(toDisplay ad: MAAd, withError error: MAError) {
        methodChannel.invokeMethod("onAdFailed", arguments: error.message)
    }

    func didClick(_ ad: MAAd) {
        methodChannel.invokeMethod("onAdClicked", arguments: nil)
    }

    func didHide(_ ad: MAAd) {
        methodChannel.invokeMethod("onAdClosed", arguments: nil)
    }

    func didDisplay(_ ad: MAAd) {}
    func didExpand(_ ad: MAAd) {}
    func didCollapse(_ ad: MAAd) {}
}
